pkg/ring/basic_lifecycler.go

```diff
@@ -469,20 +469,21 @@ func (l *BasicLifecycler) heartbeat(ctx context.Context) {
 // changeState of the instance within the ring. This function is guaranteed
 // to be called within the lifecycler main goroutine.
 func (l *BasicLifecycler) changeState(ctx context.Context, state InstanceState) error {
+	oldState := l.GetState()
 	err := l.updateInstance(ctx, func(_ *Desc, i *InstanceDesc) bool {
 		// No-op if the state hasn't changed.
 		if i.State == state {
 			return false
 		}
-
 		i.State = state
 		return true
 	})
 
 	if err != nil {
-		level.Warn(l.logger).Log("msg", "failed to change instance state in the ring", "from", l.GetState(), "to", state, "err", err)
+		level.Info(l.logger).Log("msg", "failed to change instance state in the ring", "from", oldState, "to", state, "ring", l.ringName, "err", err)
+	} else {
+		level.Info(l.logger).Log("msg", "successfully changed instance state from", "old_state", oldState, "to", state, "ring", l.ringName)
```

Comment:
Judging by this [line of code](https://github.com/wenxu1024/cortex/blob/cbcf0390f01d41220f32b379b4bd5376d8d370cf/pkg/ring/basic_lifecycler.go#L426), `oldState` and `newState` may be the same, which makes me wonder if we will have too much logs by logging successful change?

Reply:
good point. I'll remove the successful log.

Reply:
The changeState seems only called when there is a real state change. I didn't see it get called in the heartbeat, so the frequency of the log should not be very large I guess.
The linked code is updateInstance which do get called in the heartbeat and logs there could be amplified. But here the log is added outside the updateInstance function, so we are okay, I think.